### DISCUSSION TITLE
Could anyone give clarity on port direction?

### DISCUSSION THREAD

**Discussion by JackNWhite** [nasa/fprime#1401] [Q&A] [answered]
Hi,

The F' docs say:

> When using a port, the user specifies directionality. It can represent an input, or an output. This direction pertains to the direction of invocation from one originating Component to another and not necessarily the direction of data flow i.e. output ports (invoking out to another Component) can in some cases retrieve data.

This confuses me. You can invoke something, but you cannot "invoke out to" anything.

If I want to send data out of my component, I invoke my component's outport port, no? And then the autocoded stuff sends that to an input port?

But that *is* the direction of data flow. And surely, my component doesn't invoke the ports of other components directly, because that's the tail wagging the dog. Presumably, if I want data from another component, I again, invoke my own output port, which causes the autocoded code to invoke the input pot on the other the component. Then that component retrieves the data and sends it to its own output port, which calls my component's input port.

How could port direction mean anything other than this?

**Comment by timcanham** [comment, marked as answer]
Perhaps the wording "invoke out to" is a little awkward. The direction of the port is the direction of the invocation of the interface of the port, which really boils down to a function call. The component calls its output port, which calls the input port on the destination component, which invokes the handlers in the component. The reason we say the direction of the invocation is not necessarily the direction of the data flow is because ports can return data under certain conditions (sync or guarded ports on the destination component). You can get data from a destination component two ways then:

1) Request data from the component, then the component returns data via its own output port to one of your input ports. That's the pattern you described. This can make the calling component's logic more complicated because it has to track the return port call in another handler. A good example of this is the `Svc/CmdDispatcher` component. It sends commands to components, and then waits for the component to send command status back via its own output ports.

2) You can define a port to return data or modify a parameter passed by reference. This would be invocation out, data flow in.
